### PR TITLE
Return filenames on save

### PR DESCRIPTION
**Pull request by vferat**
#### Reference issue
Implements #12809.

#### What does this implement/fix?
Return filenames when saving files which might be splitted. Users can have access to the path of each split without additional code.

### CODE REVIEW

File: doc/changes/v1.8.rst [hunk 1]
```diff
@@ -129,7 +129,7 @@ New features
 - Add support for storing Fourier coefficients in :class:`mne.time_frequency.Spectrum`,
   :class:`mne.time_frequency.EpochsSpectrum`, :class:`mne.time_frequency.SpectrumArray`,
   and :class:`mne.time_frequency.EpochsSpectrumArray` objects, by `Thomas Binns`_. (`#12747 <https://github.com/mne-tools/mne-python/pulls/12747>`__)
-
+- :meth:`~mne.io.Raw` and :meth:`~mne.Epochs.save` now return the path to the saved file(s), by `Victor Ferat`_. (`#12811 <https://github.com/mne-tools/mne-python/pull/12811>`__)
```
Comment (mscheltienne) [inline doc/changes/v1.8.rst:132]: The changelog system has changed. The file you modified is automatically generated on release (1.8 is the current `stable` version). Instead, create a file `12811.newfeature.rst` in `doc/changes/devel`.

File: mne/epochs.py [hunk 2]
```diff
@@ -2195,6 +2195,12 @@ def save(
             .. versionadded:: 0.24
         %(verbose)s
 
+        Returns
+        -------
+        fnames : List of path-like
+            List of path-like objects containing the path to each file split.
+            .. versionadded:: 0.18.1
```
Comment (mscheltienne) [inline mne/epochs.py:2202]:
```suggestion

            .. versionadded:: 1.9
```

File: mne/io/base.py [hunk 3]
```diff
@@ -1663,6 +1663,12 @@ def save(
             .. versionadded:: 0.17
         %(verbose)s
 
+        Returns
+        -------
+        fnames : List of path-like
+            List of path-like objects containing the path to each file split.
+            .. versionadded:: 0.18.1
```
Comment (mscheltienne) [inline mne/io/base.py:1670]:
```suggestion

            .. versionadded:: 1.9
```

**Review by mscheltienne — COMMENTED** [review]
Are the tests modified generating a single file or split-files?
It would be nice to test both, thus if a single-file is generated, could you also add a tests which saves a short `raw`/`epochs` (can be generated from `numpy` array) with a small `split_size` argument?

**Comment by vferat** [comment]
While running tests on my computer, the  `ant\tests\test_ant.py` tests are not skipped and raise:
```
FileNotFoundError: fname does not exist: "C:\Users\victor.ferat\Documents\GitHub\mnepython\mne\io\antio\CA_208\test_CA_208.cnt"
```
Some  `@testing.requires_testing_data` might be missing here

**Comment by vferat** [comment]
For your info, even if all tests are green here, I still have some failed tests on my local **windows** machine:
```
FAILED fiff\tests\test_raw_fiff.py::test_file_like[False-str-file] - OSError: [Errno 22] Invalid argument: 'C:\\Users\\victor.ferat\\AppData\\Local\\Temp\\pytest-of-victor.ferat\\pytest-16\\test_file_like_False_str_file_0\\memmap'
FAILED fiff\tests\test_raw_fiff.py::test_file_like[False-str-bytes] - OSError: [Errno 22] Invalid argument: 'C:\\Users\\victor.ferat\\AppData\\Local\\Temp\\pytest-of-victor.ferat\\pytest-16\\test_file_like_False_str_bytes0\\memmap'
FAILED fiff\tests\test_raw_fiff.py::test_file_like[True-str-file] - OSError: [Errno 22] Invalid argument: 'C:\\Users\\victor.ferat\\AppData\\Local\\Temp\\pytest-of-victor.ferat\\pytest-16\\test_file_like_True_str_file_0\\memmap'
FAILED fiff\tests\test_raw_fiff.py::test_file_like[True-str-bytes] - OSError: [Errno 22] Invalid argument: 'C:\\Users\\victor.ferat\\AppData\\Local\\Temp\\pytest-of-victor.ferat\\pytest-16\\test_file_like_True_str_bytes_0\\memmap'
```

```
_________________________________________________________________________________________________________________________________________________
fiff\tests\test_raw_fiff.py:2061: in test_file_like
    _test_raw_reader(read_raw_fif, **kwargs)
tests\test_raw.py:486: in _test_raw_reader
    _test_raw_crop(reader, t_prop, kwargs)
tests\test_raw.py:528: in _test_raw_crop
    raw_1 = reader(**kwargs)
fiff\raw.py:532: in read_raw_fif
    return Raw(
<decorator-gen-357>:12: in __init__
    ???
fiff\raw.py:154: in __init__
    self._preload_data(preload)
base.py:596: in _preload_data
    self._data = self._read_segment(data_buffer=data_buffer)
<decorator-gen-340>:12: in _read_segment
    ???
base.py:415: in _read_segment
    data = _allocate_data(data_buffer, data_shape, dtype)
base.py:2475: in _allocate_data
    data = np.memmap(str(preload), mode="w+", dtype=dtype, shape=shape)
..\..\..\..\..\AppData\Local\miniconda3\envs\mne_dev\Lib\site-packages\numpy\_core\memmap.py:233: in __new__
    f_ctx = open(
E   OSError: [Errno 22] Invalid argument: 'C:\\Users\\victor.ferat\\AppData\\Local\\Temp\\pytest-of-victor.ferat\\pytest-16\\test_file_like_True_
```

The errors also raise on the `main` branch, so this might not be related to the current changes, but might be related to https://github.com/numpy/numpy/issues/25665

File: mne/io/tests/test_raw.py [hunk 4]
```diff
@@ -344,8 +344,11 @@ def _test_raw_reader(
     # Test saving and reading
     out_fname = op.join(tempdir, "test_raw.fif")
     raw = concatenate_raws([raw])
-    raw.save(out_fname, tmax=raw.times[-1], overwrite=True, buffer_size_sec=1)
-
+    filenames = raw.save(
+        out_fname, tmax=raw.times[-1], overwrite=True, buffer_size_sec=1
+    )
+    for filename in filenames:
+        assert op.isfile(filename)
```
Comment (larsoner) [inline mne/io/tests/test_raw.py:351]:
```suggestion
        assert filename.is_file()
```

File: mne/tests/test_epochs.py [hunk 5]
```diff
@@ -1690,12 +1690,13 @@ def test_split_naming(
...
         dst_fpath.parent.mkdir(parents=True)
 
-    epochs.save(dst_fpath, verbose=True, **save_kwargs)
+    split_fnames = epochs.save(dst_fpath, verbose=True, **save_kwargs)
 
     # check that the filenames match the intended pattern
     assert len(list(dst_fpath.parent.iterdir())) == n_files
     assert not (tmp_path / split_fname_fn(n_files)).is_file()
     want_paths = [tmp_path / split_fname_fn(i) for i in range(n_files)]
+    assert np.all(split_fnames == want_paths)
```
Comment (larsoner) [inline mne/tests/test_epochs.py:1699]: `split_fnames` should be a list so this should work without any `all`
```suggestion
    assert split_fnames == want_paths
```

File: mne/tests/test_epochs.py [hunk 6]
```diff
@@ -1750,7 +1751,9 @@ def test_saved_fname_no_splitting(
     dst_fpath = tmp_path / dst_fname
     split_1_fpath = tmp_path / split_1_fname
 
-    epochs.save(dst_fpath, split_naming=split_naming, verbose=True)
+    filenames = epochs.save(dst_fpath, split_naming=split_naming, verbose=True)
+    assert len(filenames) == 1
+    assert filenames[0] == Path(dst_fpath)
```
Comment (larsoner) [inline mne/tests/test_epochs.py:1756]: Already a path so no need to cast, and can just check both length and content simultaneously
```suggestion
    assert filenames == [dst_fpath]
```

**Review by larsoner — COMMENTED** [review]
Just some minor comments/ideas otherwise looks good and seems like CIs are (soon to be) happy!

File: mne/epochs.py [hunk 7]
```diff
@@ -2195,6 +2195,12 @@ def save(
             .. versionadded:: 0.24
         %(verbose)s
 
+        Returns
+        -------
+        fnames : List of path-like
+            List of path-like objects containing the path to each file split.
+            .. versionadded:: 0.19
```
Comment (mscheltienne) [inline mne/epochs.py:2202]:
```suggestion
            .. versionadded:: 1.9
```

File: mne/io/base.py [hunk 8]
```diff
@@ -1663,6 +1663,12 @@ def save(
             .. versionadded:: 0.17
         %(verbose)s
 
+        Returns
+        -------
+        fnames : List of path-like
+            List of path-like objects containing the path to each file split.
+            .. versionadded:: 0.19
```
Comment (mscheltienne) [inline mne/io/base.py:1670]:
```suggestion
            .. versionadded:: 1.9
```

**Review by mscheltienne — APPROVED** [review]
Aside from the remaining minor comments, LGTM! Thanks @vferat!

**Comment by mscheltienne** [comment]
FYI, if you go to the "Files changed" tab, you can "Add a suggestion to batch" and then commit all suggestion at once ;)

If you are happy with it, you can remove the draft status and I'll mark it for merge when green. Thanks!

**Pull request merged**